Allow prometheus user to run `lnetctl`

RDruon opened the pull request:
`prometheus` user do not have enough permission to get `lnetctl` statistics.

Add it to the sudoers and allows passwordless sudo execution of `lnetctl`.

ip1981 commented:
Maybe grant it that permission?

RDruon commented:
`lnetctl` requires access to `/dev/lnet` compared to `lctl`:
```
[root@ai400-001 ~]# su -l prometheus -s /usr/sbin/lctl get_param obdfilter.ai400x2-OST0001.stats
Last login: Wed Aug  2 08:59:26 UTC 2023 on pts/0
obdfilter.ai400x2-OST0001.stats=
snapshot_time             1690966809.345507985 secs.nsecs
start_time                1690966238.538483962 secs.nsecs
elapsed_time              570.807024023 secs.nsecs
create                    8 samples [usecs] 1 6134 20703 108412567
statfs                    434 samples [usecs] 0 21 1308 8436
get_info                  4 samples [usecs] 646 1955 5639 8888911
[root@ai400-001 ~]# su -l prometheus -s /usr/sbin/lnetctl net show
Last login: Wed Aug  2 09:00:09 UTC 2023 on pts/0
opening /dev/lnet failed: Permission denied
hint: the kernel modules may not be loaded
show:
    - net:
          errno: -13
          descr: "cannot get networks: Permission denied"
[root@ai400-001 ~]# ls -lah /dev/lnet
crw------- 1 root root 10, 60 Aug  2 08:47 /dev/lnet
```

How would you solve this?

ip1981 commented on sudoers_file/prometheus:
```diff
@@ -0,0 +1 @@
+prometheus ALL=(ALL) NOPASSWD: /usr/sbin/lnetctl
```
I think this will be more specific:
```suggestion
prometheus ALL=(ALL) NOPASSWD: /usr/sbin/lnetctl stats show
prometheus ALL=(ALL) NOPASSWD: /usr/sbin/lnetctl net show -v 4
```

ip1981 commented on src/main.rs:
```diff
@@ -46,8 +46,8 @@ async fn main() {
-        let lnetctl_stats_output = Command::new("lnetctl")
-            .args(["stats", "show"])
+        let lnetctl_stats_output = Command::new("sudo")
+            .args(["lnetctl", "stats", "show"])
```
```suggestion
            .args(["/usr/sbin/lnetctl", "stats", "show"])
```

ip1981 commented on src/main.rs:
```diff
@@ -36,8 +36,8 @@ async fn main() {
-        let lnetctl = Command::new("lnetctl")
-            .args(["net", "show", "-v", "4"])
+        let lnetctl = Command::new("sudo")
+            .args(["lnetctl", "net", "show", "-v", "4"])
```
```suggestion
            .args(["/usr/sbin/lnetctl", "net", "show", "-v", "4"])
```

ip1981 reviewed (approved):
Ok, if it works :)

jgrund commented:
@RDruon Tested?

ip1981 commented:
You can test just with the sudoers file running commands as prometheus.

RDruon commented:
> @RDruon Tested?

Yes. 

```
[root@ai400-001 ~]# systemctl status lustrefs_exporter
● lustrefs_exporter.service - Prometheus exporter for Lustre filesystem
   Loaded: loaded (/usr/lib/systemd/system/lustrefs_exporter.service; disabled; vendor preset: disabled)
   Active: active (running) since Wed 2023-08-02 14:35:36 UTC; 55min ago
     Docs: https://github.com/whamcloud/lustrefs-exporter
 Main PID: 502108 (lustrefs_export)
    Tasks: 21 (limit: 960144)
   Memory: 2.4M
   CGroup: /system.slice/lustrefs_exporter.service
           └─502108 /usr/bin/lustrefs_exporter

Aug 02 15:30:36 ai400-001 sudo[638511]: prometheus : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/sbin/lnetctl stats show
Aug 02 15:30:46 ai400-001 lustrefs_exporter[502108]: in our render_prometheus(request == Request { method: GET, uri: /metrics, version: HTTP/1.1, headers: {"host": "10.36.5.190:32221", "user-agent": "Prometheus/2.46.0", "accept": "application/openmetrics-text;version=1.0.0,appli>
Aug 02 15:30:46 ai400-001 sudo[638959]: prometheus : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/sbin/lnetctl net show -v 4
Aug 02 15:30:46 ai400-001 sudo[638962]: prometheus : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/sbin/lnetctl stats show
Aug 02 15:30:56 ai400-001 lustrefs_exporter[502108]: in our render_prometheus(request == Request { method: GET, uri: /metrics, version: HTTP/1.1, headers: {"host": "10.36.5.190:32221", "user-agent": "Prometheus/2.46.0", "accept": "application/openmetrics-text;version=1.0.0,appli>
Aug 02 15:30:56 ai400-001 sudo[639246]: prometheus : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/sbin/lnetctl net show -v 4
Aug 02 15:30:56 ai400-001 sudo[639249]: prometheus : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/sbin/lnetctl stats show
Aug 02 15:31:06 ai400-001 lustrefs_exporter[502108]: in our render_prometheus(request == Request { method: GET, uri: /metrics, version: HTTP/1.1, headers: {"host": "10.36.5.190:32221", "user-agent": "Prometheus/2.46.0", "accept": "application/openmetrics-text;version=1.0.0,appli>
Aug 02 15:31:06 ai400-001 sudo[639529]: prometheus : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/sbin/lnetctl net show -v 4
Aug 02 15:31:06 ai400-001 sudo[639532]: prometheus : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/sbin/lnetctl stats show
[root@ai400-001 ~]# cat /etc/sudoers.d/prometheus
prometheus ALL=(ALL) NOPASSWD: /usr/sbin/lnetctl net show -v 4, /usr/sbin/lnetctl stats show
[root@ai400-001 ~]# curl -s http://localhost:32221/metrics | grep lustre_send_bytes_total
# HELP lustre_send_bytes_total Total number of bytes that have been sent
# TYPE lustre_send_bytes_total counter
lustre_send_bytes_total 25404816 1690990294501
[root@ai400-001 ~]# curl -s http://localhost:32221/metrics | grep lustre_send_count_total
# HELP lustre_send_count_total Total number of messages that have been sent
# TYPE lustre_send_count_total counter
lustre_send_count_total{nid="0@lo"} 16627 1690990301588
lustre_send_count_total{nid="172.16.232.190@o2ib"} 70542 1690990301588
```

ip1981 commented:
Does sudo allow running slightly different commands? It also should print something in logs.

RDruon commented:
> Does sudo allow running slightly different commands (it should not)? It also should print something in logs.

Logs are shown in the journalctl logs.:
```
Aug 02 15:31:06 ai400-001 lustrefs_exporter[502108]: in our render_prometheus(request == Request { method: GET, uri: /metrics, version: HTTP/1.1, headers: {"host": "10.36.5.190:32221", "user-agent": "Prometheus/2.46.0", "accept": "application/openmetrics-text;version=1.0.0,appli>
Aug 02 15:31:06 ai400-001 sudo[639529]: prometheus : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/sbin/lnetctl net show -v 4
Aug 02 15:31:06 ai400-001 sudo[639532]: prometheus : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/sbin/lnetctl stats show
```

Not sure how to test with `prometheus` and `sudo`:
```
[root@ai400-001 ~]# su - prometheus
Last login: Wed Aug  2 15:34:42 UTC 2023 on pts/0
This account is currently not available.
```
but
```
[root@ai400-001 ~]# sudo -l -U prometheus
Matching Defaults entries for prometheus on ai400-001:
    !visiblepw, always_set_home, match_group_by_gid, always_query_group_plugin, env_reset, env_keep="COLORS DISPLAY HOSTNAME HISTSIZE KDEDIR LS_COLORS", env_keep+="MAIL PS1 PS2 QTDIR USERNAME LANG LC_ADDRESS LC_CTYPE", env_keep+="LC_COLLATE LC_IDENTIFICATION LC_MEASUREMENT
    LC_MESSAGES", env_keep+="LC_MONETARY LC_NAME LC_NUMERIC LC_PAPER LC_TELEPHONE", env_keep+="LC_TIME LC_ALL LANGUAGE LINGUAS _XKB_CHARSET XAUTHORITY", secure_path=/sbin\:/bin\:/usr/sbin\:/usr/bin

User prometheus may run the following commands on ai400-001:
    (ALL) NOPASSWD: /usr/sbin/lnetctl net show -v 4, /usr/sbin/lnetctl stats show
```

ip1981 commented:
Ok. `Something like sudo -u prometheus /usr/sbin/lnetctl set...` should be refused :)

RDruon commented:
> Ok. Something like `sudo -u prometheus /usr/sbin/lnetctl set...` should be refused :)
> 
> Anyway, it's me being paranoid. Everything looks good.

It's a bit weird when using `sudo -u prometheus`:
```
[root@ai400-001 ~]# sudo -u prometheus /usr/sbin/lnetctl route show
opening /dev/lnet failed: Permission denied
hint: the kernel modules may not be loaded
show:
    - route:
          errno: -13
          descr: "cannot get routes: Permission denied"
[root@ai400-001 ~]# sudo -u prometheus /usr/sbin/lnetctl net show -v 4
opening /dev/lnet failed: Permission denied
hint: the kernel modules may not be loaded
show:
    - net:
          errno: -13
          descr: "cannot get networks: Permission denied"
```
It allows me to run the commands where it should, but it seems it's not doing `sudo $CMD` as `sudo /usr/sbin/lnetctl net show -v 4` does work in code

RDruon commented:
> Ok. Something like `sudo -u prometheus /usr/sbin/lnetctl set...` should be refused :)
> 
> Anyway, it's me being paranoid. Everything looks good.

I was missing the double `sudo`:
```
[root@ai400-001 ~]# sudo -u prometheus sudo /usr/sbin/lnetctl net show -v 4
net:
    - net type: lo
      local NI(s):
        - nid: 0@lo
          status: up
          statistics:
              send_count: 17701
              recv_count: 17160
              drop_count: 541
          sent_stats:
              put: 17701
              get: 0
              reply: 0
              ack: 0
              hello: 0
          received_stats:
              put: 17149
              get: 0
              reply: 0
              ack: 11
              hello: 0
          dropped_stats:
              put: 541
              get: 0
              reply: 0
              ack: 0
              hello: 0
          health stats:
              fatal_error: 0
              health value: 1000
              interrupts: 0
              dropped: 0
              aborted: 0
              no route: 0
              timeouts: 0
              error: 0
          tunables:
              peer_timeout: 0
              peer_credits: 0
              peer_buffer_credits: 0
              credits: 0
          lnd tunables:
          dev cpt: 0
          CPT: "[0,1,2,3,4]"
    - net type: o2ib
      local NI(s):
        - nid: 172.16.232.190@o2ib
          status: up
          interfaces:
              0: mlxib0
          statistics:
              send_count: 75319
              recv_count: 74782
              drop_count: 28
          sent_stats:
              put: 63980
              get: 11339
              reply: 0
              ack: 0
              hello: 0
          received_stats:
              put: 63416
              get: 5671
              reply: 5668
              ack: 27
              hello: 0
          dropped_stats:
              put: 28
              get: 0
              reply: 0
              ack: 0
              hello: 0
          health stats:
              fatal_error: 0
              health value: 1000
              interrupts: 0
              dropped: 0
              aborted: 0
              no route: 0
              timeouts: 0
              error: 0
          tunables:
              peer_timeout: 180
              peer_credits: 32
              peer_buffer_credits: 0
              credits: 256
          lnd tunables:
              peercredits_hiw: 16
              map_on_demand: 1
              concurrent_sends: 64
              fmr_pool_size: 512
              fmr_flush_trigger: 384
              fmr_cache: 1
              ntx: 512
              conns_per_peer: 1
          dev cpt: 0
          CPT: "[0,1,2,3,4]"
[root@ai400-001 ~]# sudo -u prometheus sudo /usr/sbin/lnetctl net show -v 3

We trust you have received the usual lecture from the local System
Administrator. It usually boils down to these three things:

    #1) Respect the privacy of others.
    #2) Think before you type.
    #3) With great power comes great responsibility.
```

jgrund commented:
@RDruon This needs a followup with a new release
